trpcultivate_phenotypes/src/Plugin/Validators/validDataFile.php

```diff
@@ -0,0 +1,197 @@
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container, array $configuration, $plugin_id, $plugin_definition){
+    return new static(
+      $configuration,
+      $plugin_id,
+      $plugin_definition,
+      $container->get('entity_type.manager')
+    );
+  }
+
+  /**
+   * Perform validation of data file.
+   * Checks include:
+   *  - Parameter filename or file id is valid.
+   *  - Has Drupal File Id number assigned and can be loaded.
+   *  - Valid file extension and file mime type configured by the file importer instance.
+   *  - File exists and not empty.
+   *  - File can be opened.
+   * 
+   * @param string $filename
+   *   The location of a file within the file system. 
+   * @param integer $fid
+   *   The unique identifier (fid) of a file that is managed by Drupal File System.    
+   * 
+   * @return array
+   *   An associative array with the following keys.
+   *    - case: a developer focused string describing the case checked.
+   *    - valid: either TRUE or FALSE depending on if the file is valid or not.
+   *    - failedItems: an array of "items" that failed to be used in the message to the user. This is an empty array if the file input was valid.
+   */
+  public function validateFile($filename, $fid = NULL) {
+    // Validator response values for a valid file.
+    $case = 'Data file is valid';
+    $valid = TRUE;
+    $failed_items = [];
+
+    // Parameter check, verify that the filename/file path is valid.
+    if (empty($filename) && is_null($fid)) {
+      return [
+        'case' => 'Filename is empty',
+        'valid' => FALSE,
+        'failedItems' => ['filename' => $filename]
+      ];
+    }
+    
+    // Parameter check, verify the file id number is not 0 or negative values.
+    if (!is_null($fid) && $fid <= 0) {
+      return [
+        'case' => 'Invalid file id number',
+        'valid' => FALSE,
+        'failedItems' => ['file_id' => $fid]
+      ];
+    } 
+
+    // File input.
+    $file_input = (is_null($fid)) ? $filename : $fid;
+    
+    // Load file object.
+    if (is_numeric($file_input)) {
+      // The file input is integer value, the file id number.
+      // Load the file object by fid number.
+      $file_id = $fid;
+      $file_object = File::load($file_id);
+
+      $input_key = 'file_id';
+    }
+    else {
+      // The file input is a string value, a path to the file.
+      // Locate the file entity by uri and load the file object using
+      // the returned file id number that matched.
+      $file_entities = $this->service_EntityTypeManager
+        ->getStorage('file')
+        ->loadByProperties(['uri' => $filename]);
+      
+      $file_entity = reset($file_entities);
+      $file_object = 0;
+      
+      if ($file_entity) {
+        $file_id = $file_entity->get('fid')->value;
+        $file_object = File::load($file_id);
+      }
+
+      $input_key = 'filename';
+    }
+    
+
+    // Validate data file:
+
+    if ($file_object) {
```

Comment:
I would inverse all of these if statements so that you can set the validation array inside the if block instead of in the else. Currently, it sadly makes it very hard to follow :(

Reply:
I think it would be even more readable to just keep each check in it's own if block and return when there is a failure.

Something like:
```php
// Check that the file is not blank by inspecting the file size
// to see if it is greater than 0.
$file_size = $file_object->getSize();
if (!is_numeric($file_size) OR ($file_size <= 0)) {
  //return the failure.
}

// Check that both the file extension and file MIME type
// are supported by the importer.
$file_mime_type = $file_object->getMimeType();
$file_filename  = $file_object->getFileName();
$file_extension = pathinfo($file_filename, PATHINFO_EXTENSION);
```

Note that there is no nesting of if statements this way because it is not needed. We will only get to the second check if the first one passes.

Reply:
My concerns for this file are resolved ✅ but it would be good to see the same approach in  ValidDataFile and your other validators in general to improve readability.